drivers: video: Add SMH option for video buffer

LucasTambor opened the pull request:
This commit enables the user to choose whether to
allocate the video buffer from the video heap pool
or use a memory region with specific extra capabilities,
such as being cacheable/non-cacheable or using external
memory.

uLipe commented on drivers/video/Kconfig:
```diff
@@ -31,6 +31,21 @@ config VIDEO_BUFFER_POOL_ALIGN
 	default 64
 
+config VIDEO_BUFFER_USE_SHARED_MULTI_HEAP
+	bool "Use shared multi heap for video buffer"
```
I think this option would depend on the presence of an external memory device,  I'm wondering if would be good to cheate a `chosen` property on the device tree to indicate that presence and use it to drive this option.

for example:

```
{
  chosen, zephyr-ext-ram = &ext-ram0;
}

```

LucasTambor replied:
I'm not sure. #76980 extends SMH attributes to include external memory as a possible heap capability. This feature provides more capabilities that don't require an external memory.

uLipe replied:
That makes sense.

LucasTambor committed:
drivers: video: Add SMH option for video buffer

This commit enables the user to choose whether to
allocate the video buffer from the video heap pool
or use a memory region with specific extra capabilities,
such as being cacheable/non-cacheable or using external
memory.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

epc-ake commented:
I was able to run it on the `xiao_esp32s3` together with #75331 the changes from #74974 and the patch you send me on discord.  :+1: 
However, I was not able to go above a resolution of `352x288 RGB565` before I run into `<err> esp32_cam: DMA error: -32`. 
Which seems to be related with dma descriptors somehow... 😓
Also the issue with the generated overhead I mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/74974#issuecomment-2230305434) caused trough the dma_api really comes into play with higher video resolutions...

However, these issues most likely have nothing to do with this PR. So I'd call the test successful and try to fix the issues in the DMA and Video driver.

LucasTambor commented:
@epc-ake The `DMA error: -32` is related to the RX Watermark. Can you try increase `gdma_ll_rx_set_water_mark` and see if it helps?

epc-ake commented:
> @epc-ake The `DMA error: -32` is related to the RX Watermark. Can you try increase `gdma_ll_rx_set_water_mark` and see if it helps?

Yes, this seems to fix it! 
Can you maybe explain to me what `gdma_ll_rx_set_water_mark` does? Its not about the size of the dmas fifo, is it?
I needed to set it to `65`. A value of `64` and lower didn't work and evan crashes the system  with `<err> os:  ** CPU 0 EXCCAUSE 9 (load/store alignment)`...
Now I'm able to capture the full `1600x1200 RGBA` resolution 🚀

uLipe reviewed (dismissed):
Besides CI checks, LGTM

loicpoulain commented:
Just a question, is it something the 'user' should be able to configure, or is it something depending on the platform/IP requirements?

LucasTambor commented:
> Just a question, is it something the 'user' should be able to configure, or is it something depending on the platform/IP requirements?

It's up to the user to use to configure it or not. If their platform supports it and they want to use a memory region with capabilities such as external/cacheable/non-cacheable (depends on what SMH supports) is now possible, otherwise if they do nothing it will use default sys_heap as it used to be. Did I answer it? haha

josuah commented on drivers/video/Kconfig:
```diff
@@ -31,6 +31,21 @@ config VIDEO_BUFFER_POOL_ALIGN
+	bool "Use shared multi heap for video buffer"
+	default n
+
+config VIDEO_BUFFER_SMH_ATTRIBUTE
+	int "Shared multi heap attribute for video buffer"
+	depends on VIDEO_BUFFER_USE_SHARED_MULTI_HEAP
+	default 0
+	range 0 2
+	help
+	  Shared multi heap attribute for video buffer:
+	  0: SMH_REG_ATTR_CACHEABLE
+	  1: SMH_REG_ATTR_NON_CACHEABLE
+	  2: SMH_REG_ATTR_EXTERNAL
```
It is also possible to make use of *choice* depending on tastes:
https://docs.zephyrproject.org/latest/build/kconfig/setting.html#configuring-choices

It is also possible to make them all optional, so that `VIDEO_BUFFER_USE_SHARED_MULTI_HEAP` and `VIDEO_BUFFER_SMH_ATTRIBUTE` can be merged together:
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#optional-choices

Not sure what is the best as I am still learning Zephyr best practices, maybe how you proposed is best!

LucasTambor replied:
My idea here was to use `CONFIG_VIDEO_BUFFER_SMH_ATTRIBUTE` directly as an argument in `shared_multi_heap_aligned_alloc` based on the `SMH` enum:

```
enum shared_multi_heap_attr {
	/** cacheable */
	SMH_REG_ATTR_CACHEABLE,

	/** non-cacheable */
	SMH_REG_ATTR_NON_CACHEABLE,

	/** external Memory */
	SMH_REG_ATTR_EXTERNAL,

	/** must be the last item */
	SMH_REG_ATTR_NUM,
};
```

I think this might be cleaner than to have choices in Kconfig and have to use `#ifdefs`. What do you think?

josuah replied:
What you picked is the least bureaucratic! Thank you for clarifying, this also teaches me how to map Kconfig to C enums.

LucasTambor commented:
@andyross @npitre
